malcolm/core/task.py

```diff
@@ -0,0 +1,240 @@
+        return f
+
+    def post_asynch(self, method, params):
+        """Asynchronously calls a function on a child block
+
+            Returns a list of one future which will proved the return value
+            on completion"""
+        endpoint = [method.parent.name, method.name]
+        request = Request.Post(None, self.q, endpoint, params)
+        f = Future(self)
+        new_id = self._save_future(f)
+        request.set_id(new_id)
+        self.process.q.put(request)
+
+        return [f]
+
+    def subscribe(self, attr, callback, *args):
+        """subscribe to changes in a given attribute and call callback
+            with (*args) when it changes
+
+            Returns an id for the subscription"""
+
+        endpoint = [attr.parent.name, attr.name]
+        request = Request.Subscribe(None, self.q, endpoint, False)
+        new_id = self._save_subscription(endpoint, callback, *args)
+        request.set_id(new_id)
+        self.process.q.put(request)
+
+        return new_id
+
+    def unsubscribe(self, id_):
+        """Terminates the subscription to an attribute
+
+            Args:
+                id_ (Int): the identifier of the subscription to halt"""
+
+        # todo - Request does not yet have Unsubscribe
+        # (endpoint, func, args) = self._subscriptions[id_]
+        # request = Request.Unsubscribe(None, self.q, endpoint)
+        old_subscription = self._subscriptions.pop(id_)
+        pass
+
+    def stop(self):
+        """Puts an abort on the queue"""
+        self.q.put(Task.TASK_STOP)
+
+    def wait_all(self, futures, timeout=None):
+        """services all futures until the list 'futures' are all done
+            then returns. Calls relevant subscription callbacks as they
+            come off the queue and raises an exception on abort
+
+            Args:
+                futures ([Future] or Future): a future or list of all futures
+                    that the caller wants to wait for
+                timeout (Float) time in seconds to wait for responses, wait
+                    forever if None"""
+        if not isinstance(futures, list):
+            futures = [futures]
+
+        futures = [f for f in futures if not f.done()]
+
+        while futures:
+            self.log_debug("wait_all awaiting response ...")
+            response = self.q.get(True, timeout)
+            self.log_debug("wait_all received response %s", response)
+            if response is Task.TASK_STOP:
+                raise RuntimeWarning("Task aborted")
+            elif response.id_ in self._futures:
+                f = self._update_future(response)
+                if f in futures:
+                    futures.remove(f)
+            elif response.id_ in self._subscriptions:
+                result = self._invoke_callback(response)
+                # Task's _match_update callback returns a future if it has
+                #  filled it (other callbacks not expected to return a val)
+                if result in futures:
```

Comment:
Should put a guard somewhere to only do this if callback function is self._match_update